### PR TITLE
Ansible for offline installer

### PR DESCRIPTION
**Pull request by jr0d**
Isolated PR containing only Ansible changes for airgapped bundle support.

### CODE REVIEW

File: ansible/roles/offline/tasks/upload.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+- name: create offline directory
+  file:
+    path: "{{ offline.remote_bundle_os_packages_path }}"
+    state: directory
+
+- name: upload and extract offline bundle
+  unarchive:
+    copy: true
+    src: "{{ offline.local_bundle_path }}"
+    dest: "{{ offline.remote_bundle_os_packages_path }}"
```
Comment (dkoshkin) [inline ansible/roles/offline/tasks/upload.yaml:14]: Nice, we can do something similar for the images.

**Review by dkoshkin — COMMENTED** [review]
For tasks like 
```
- name: install kubectl rpm package
  yum:
    name: "{{ 'kubectl-' + package_versions.kubernetes_rpm }}"
    state: present
    update_cache: yes
  register: result
  until: result is success
  retries: 3
  delay: 3
```
And other `yum` tasks does it not try to reach out to the internet update the cache?

Could we use `enablerepo/disablerepo ` dynamically `when: offline_path.stat.isdir is defined and offline_path.stat.isdir` to and just rely on the existing tasks to install all these packages?

File: ansible/roles/packages/tasks/install-redhat-offline.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
+---
+- name: upload offline repository
+  copy:
+    src: offline.repo
+    dest: /etc/yum.repos.d/offline.repo
+
+- name: install offline packages
```
Comment (dkoshkin) [inline ansible/roles/packages/tasks/install-redhat-offline.yaml:7]: We should call out the limitation about air-gapped support and `nvidia`.

Comment (dkoshkin) [reply ansible/roles/packages/tasks/install-redhat-offline.yaml:7]: Do we also need `cloud-init` packages for AWS?

**Review by dkoshkin — APPROVED** [review]
🎉

File: ansible/group_vars/all/defaults.yaml [hunk 3]
```diff
@@ -45,3 +45,6 @@ kubernetes_bins:
   - kubelet
 
 sysprep: true
+offline_local_bundle_path: ""
+offline_remote_bundle_dir: "/root/.konvoy/offline"
```
Comment (dlipovetsky) [inline ansible/group_vars/all/defaults.yaml:49]: From standup: Let's think a little bit more about the naming scheme.

There might be conflict between bundles...

(Offline introduces yet another term, can we reuse an existing one?)

**Comment by dlipovetsky** [comment]
I think we can merge this, but then iterate on it so that we have a single upload mechanism that works for both package and container image bundles. I'm thinking something like:
1. Generic task that copies everything in a local folder to a remote machine folder (`$HOME/.konvoy`)
2. Specialized tasks that extract the package and container bundles on the remote machine.

File: ansible/group_vars/all/defaults.yaml [hunk 4]
```diff
@@ -45,3 +45,6 @@ kubernetes_bins:
...
+offline_remote_bundle_file: "offline.tar.gz"
```
Comment (dlipovetsky) [inline ansible/group_vars/all/defaults.yaml:50]: Is this used anywhere? I suspect the design changed, and this is now dead code.

File: ansible/group_vars/all/defaults.yaml (hunk 3 above)
Comment (dlipovetsky) [reply ansible/group_vars/all/defaults.yaml:49]: @dkoshkin @joejulian @faiq @supershal and I talked about this, decided to use `/opt/dkp/packages` for packages (and `/opt/dkp/images` for images. Not yet sure how to change this to fit that model, though...

File: ansible/group_vars/all/system.yaml [hunk 5]
```diff
@@ -86,3 +86,10 @@ validate:
   minimum_memory_worker_mb: 2048
   # Minimum CPU cores allocated to each machine.
   minimum_cpu_cores: 2
+
+offline:
+  local_bundle_path: "{{ offline_local_bundle_path }}"
+  remote_bundle_dir: "{{ offline_remote_bundle_dir }}"
+  remote_bundle_file: "{{ offline_remote_bundle_file }}"
```
Comment (jr0d) [inline ansible/group_vars/all/system.yaml:93]: it is used here @dlipovetsky

File: ansible/group_vars/all/defaults.yaml (hunk 4 above)
Comment (jr0d) [reply ansible/group_vars/all/defaults.yaml:50]: It is used, see my next comment

File: ansible/roles/packages/tasks/main.yaml [hunk 6]
```diff
@@ -1,3 +1,14 @@
...
+  register: offline_path
+
+- name: set offline enabled fact
+  set_fact:
+    offline_mode_enabled: "{{ offline_path.stat.exists }}"
+
 - include: repo.yaml
+  when: not offline_path.stat.exists
 - include: install.yaml
+  when: not offline_path.stat.exists
```
Comment (dkoshkin) [inline ansible/roles/packages/tasks/main.yaml:14]: This should always run since the `yum` tasks have the `offline_mode_enabled` logic

**Comment by dkoshkin** [comment]
Closing in-favor of https://github.com/mesosphere/konvoy-image-builder/pull/176

**Pull request closed without merging**